### PR TITLE
[Inductor][Quant] Fix PT2E Dynamic Quant regression

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125207
* #124246
* #124041

**Summary**
Fix 2 regression issues caused by previous refactor:

- Fix the issue in dequant promotion pass with dynamic quant when the dequant node is with `tensor` overload.
- Fix numerical issue in dynamic quant, since input will convert to scales' dtype (which is `double`) to do quant operatoration with previous implementation.

**TestPlan**
```
clear && python -u -m pytest -s -v test/inductor/test_mkldnn_pattern_matcher.py -k test_dynamic_qlinear_input_dim_exceeds_2
clear && python -u -m pytest -s -v test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_dequant_promotion_dynamic_cpu
```

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -1244,16 +1244,18 @@ def quantized_decomposed_quantize_per_tensor_tensor(
...
 
     def inner_fn(idx):
-        input = input_loader(idx)
+        _input = input_loader(idx)
         _scale = scale_loader((0,) if len(scale.get_size()) == 1 else ())
         _zero_point = zero_point_loader((0,) if len(scale.get_size()) == 1 else ())
-        if scale.dtype != torch.float32:
-            _scale = ops.to_dtype(_scale, torch.float32)
+        if input.dtype != scale.dtype:
+            _input = ops.to_dtype(_input, scale.dtype)
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:1251]: Casting to double seems like a bad idea for qgpu performance, but it seems to be what eager does so I guess this is fine.

**Review by peterbell10 — APPROVED** [review]

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/lowering.py:1251]: Talked to @leslie-fang-intel offline. Let's align with eager which uses float instead of double.

Comment (leslie-fang-intel) [reply torch/_inductor/lowering.py]: In eager implementation, the scales will convert to float before computation. Before refactor, we are using double in inductor. Change back to float to align with eager.
https://github.com/pytorch/pytorch/blob/f35fe4eaf1e9fa2e631f6bf1a3eb6e5fbf14183b/torch/ao/quantization/fx/_decomposed.py#L103
https://github.com/pytorch/pytorch/blob/f35fe4eaf1e9fa2e631f6bf1a3eb6e5fbf14183b/torch/ao/quantization/fx/_decomposed.py#L37-L45

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**